oak-blob-composite/src/main/java/org/apache/jackrabbit/oak/blob/composite/CompositeDataStoreService.java

```diff
@@ -0,0 +1,180 @@
+ * under the License.
+ */
+
+package org.apache.jackrabbit.oak.blob.composite;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.io.Closer;
+import org.apache.felix.scr.annotations.Component;
+import org.apache.felix.scr.annotations.ConfigurationPolicy;
+import org.apache.felix.scr.annotations.Reference;
+import org.apache.felix.scr.annotations.ReferenceCardinality;
+import org.apache.felix.scr.annotations.ReferencePolicy;
+import org.apache.jackrabbit.core.data.DataStore;
+import org.apache.jackrabbit.core.data.DataStoreException;
+import org.apache.jackrabbit.oak.plugins.blob.datastore.AbstractDataStoreService;
+import org.apache.jackrabbit.oak.spi.blob.DataStoreProvider;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.Constants;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.service.component.ComponentContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.util.Dictionary;
+import java.util.Hashtable;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+import static org.apache.jackrabbit.oak.commons.IOUtils.closeQuietly;
+
+@Component(policy = ConfigurationPolicy.REQUIRE, name = CompositeDataStoreService.NAME)
+public class CompositeDataStoreService extends AbstractDataStoreService {
+    public static final String NAME = "org.apache.jackrabbit.oak.plugins.blob.datastore.CompositeDataStore";
+    private static Logger log = LoggerFactory.getLogger(CompositeDataStoreService.class);
+
+    private static final String DESCRIPTION = "oak.composite.datastore.description";
+
+    private CompositeDataStore dataStore = null;
+
+    @Reference(cardinality = ReferenceCardinality.OPTIONAL_MULTIPLE,
+            policy = ReferencePolicy.DYNAMIC,
+            bind = "addDelegateDataStore",
+            unbind = "removeDelegateDataStore",
+            referenceInterface = DataStoreProvider.class,
+            target="(!(service.pid=org.apache.jackrabbit.oak.plugins.blob.datastore.CompositeDataStore))"
+    )
+    private List<CompositeDataStoreDelegate> delegateDataStores = Lists.newArrayList();
+
+    private ComponentContext context;
+    private Map<String, Object> config = Maps.newConcurrentMap();
+
+    private boolean isRegistered = false;
+    private Closer closer = Closer.create();
+
+    @Override
+    protected DataStore createDataStore(ComponentContext context, Map<String, Object> config) {
+        this.context = context;
+        if (null == this.config) {
+            this.config = config;
+        }
+        else {
+            for (Map.Entry<String, Object> entry : config.entrySet()) {
+                this.config.putIfAbsent(entry.getKey(), entry.getValue());
+            }
+        }
+
+        registerCompositeDataStore();
+
+        return dataStore;
+    }
+
+    private void registerCompositeDataStore() {
+        if (delegateDataStores.isEmpty()) {
```

Comment:
IMO this is too simplistic a condition i.e it should be based on all the roles configured being satisfied with corresponding delegate datastores. Otherwise we could have a situation where for e.g. a read-only DataStore gets configured without the writable DataStore but still the CompositeDataStore is initialized but now the application would not store any blob.